### PR TITLE
Update to allow OIP txid hashes

### PR DESCRIPTION
**Pull request by blondfrogs**
- Ability to add a txhash to the asset meta data instead of an ipfs hash : This type of change allows a lot of options for integration on the ravencoin network. This is be activated the same time as the messaging RIP5

### CODE REVIEW

**Comment by blondfrogs** [comment]
#483 Also fixed in this PR

File: src/rpc/assets.cpp [hunk 1]
```diff
@@ -808,10 +817,14 @@ UniValue transfer(const JSONRPCRequest& request)
     std::string message = "";
     if (request.params.size() > 3) {
         message = request.params[3].get_str();
-        if (message.length() != 46)
+        if (message.length() != 46 || (AreMessagingDeployed() && message.length() != 64))
```
Comment (TronBlack) [inline src/rpc/assets.cpp]:
```suggestion
        if (message.length() != 46 || (AreMessagingDeployed() && message.length() != 64 && message.length() != 46))
```

File: src/rpc/assets.cpp [hunk 2]
```diff
@@ -1108,15 +1121,22 @@ UniValue reissue(const JSONRPCRequest& request)
     std::string newipfs = "";
     if (request.params.size() > 6) {
         newipfs = request.params[6].get_str();
-        if (newipfs.length() != 46)
+
+        if (newipfs.length() != 46 && (AreMessagingDeployed() && newipfs.length() != 64))
```
Comment (TronBlack) [inline src/rpc/assets.cpp]:
```suggestion
        if (newipfs.length() != 46 && (AreMessagingDeployed() && newipfs.length() != 64 && message.length() != 46))
```

File: src/qt/createassetdialog.cpp [hunk 3]
```diff
@@ -451,19 +451,19 @@ bool CreateAssetDialog::checkIPFSHash(QString hash)
...
-            showMessage("IPFS Hash must start with 'Qm'");
+            showMessage("IPFS Hash must start with 'Qm' or txid's must have size of 64 characters");
             disableCreateButton();
             return false;
         }
-        else if (hash.size() != 46) {
+        else if (hash.size() != 46 && hash.size() != 64) {
             ui->ipfsText->setStyleSheet("border: 2px solid red");
-            showMessage("IPFS Hash must have size of 46 characters");
+            showMessage("IPFS/txid Hash must have size of 46 characters");
```
Comment (TronBlack) [inline src/qt/createassetdialog.cpp]:
```suggestion
            showMessage("IPFS/txid Hash must have size of 46 characters or 64 hex characters");
```

Comment (bitspill) [inline src/qt/createassetdialog.cpp]: `IPFS/Txid` to remain consistent with other usages thus far

File: src/qt/createassetdialog.cpp [hunk 4]
```diff
@@ -451,19 +451,19 @@ bool CreateAssetDialog::checkIPFSHash(QString hash)
...
             disableCreateButton();
             return false;
-        } else if (DecodeIPFS(hash.toStdString()).empty()) {
-            showMessage("IPFS hash is not valid. Please use a valid IPFS hash");
+        } else if (DecodeAssetData(hash.toStdString()).empty()) {
+            showMessage("IPFS/txid hash is not valid. Please use a valid IPFS/txid hash");
```
Comment (bitspill) [inline src/qt/createassetdialog.cpp]: `IPFS/Txid` to remain consistent with other usages thus far

File: src/qt/reissueassetdialog.cpp [hunk 5]
```diff
@@ -638,19 +649,19 @@ bool ReissueAssetDialog::checkIPFSHash(QString hash)
...
             ui->ipfsText->setStyleSheet(STYLE_INVALID);
-            showMessage(tr("IPFS Hash must have size of 46 characters"));
+            showMessage(tr("IPFS/Txid Hash must have size of 46 characters"));
             disableReissueButton();
             return false;
-        } else if (DecodeIPFS(ui->ipfsText->text().toStdString()).empty()) {
+        } else if (DecodeAssetData(ui->ipfsText->text().toStdString()).empty()) {
             ui->ipfsText->setStyleSheet(STYLE_INVALID);
-            showMessage(tr("IPFS hash is not valid. Please use a valid IPFS hash"));
+            showMessage(tr("IPFS/Txid hash is not valid. Please use a valid IPFS hash"));
```
Comment (bitspill) [inline src/qt/reissueassetdialog.cpp]: `IPFS/Txid hash is not valid. Please use a valid IPFS or Txid hash`

File: src/qt/forms/reissueassetdialog.ui [hunk 6]
```diff
@@ -1295,7 +1305,7 @@
                  <item>
                   <widget class="QRadioButton" name="radioCustomFee">
                    <property name="text">
-                    <string>Custom:</string>
+                    <string>Cus&amp;tom:</string>
```
[the remark this answers is not in the document]
Comment (blondfrogs) [reply src/qt/forms/reissueassetdialog.ui]: This is added by Qt, and doesn't change the way the word is display. Tried removing this, but Qt puts is back in.

File: src/assets/assets.cpp [hunk 7]
```diff
@@ -3025,14 +3055,25 @@ bool CheckAmountWithUnits(const CAmount& nAmount, const int8_t nUnits)
...
 
-    return true;
+    if (AreMessagingDeployed()) {
+        if (!ret && encodedStr.length() == 64) {
+            ret = true;
+        }
+    }
+    
+    if (!ret)
+        strError = _("Invalid parameter: ipfs_hash is not valid, or txid hash is not the write length");
```
[the remark this answers is not in the document]
Comment (blondfrogs) [reply src/assets/assets.cpp]: Fixed

File: src/qt/createassetdialog.cpp (hunk 3 above)
Comment (blondfrogs) [reply src/qt/createassetdialog.cpp]: Fixed

File: src/qt/reissueassetdialog.cpp [hunk 8]
```diff
@@ -638,19 +649,19 @@ bool ReissueAssetDialog::checkIPFSHash(QString hash)
...
             ui->ipfsText->setStyleSheet(STYLE_INVALID);
-            showMessage(tr("IPFS Hash must start with 'Qm'"));
+            showMessage(tr("IPFS Hash must start with 'Qm', Txid is indicated by 0x54 0x20"));
             disableReissueButton();
             return false;
-        } else if (hash.size() != 46) {
+        } else if (hash.size() != 46 && hash.size() != 64) {
             ui->ipfsText->setStyleSheet(STYLE_INVALID);
-            showMessage(tr("IPFS Hash must have size of 46 characters"));
+            showMessage(tr("IPFS/Txid Hash must have size of 46 characters"));
```
[the remark this answers is not in the document]
Comment (blondfrogs) [reply src/qt/reissueassetdialog.cpp]: Fixed

File: src/rpc/assets.cpp [hunk 9]
```diff
@@ -471,11 +480,13 @@ UniValue getassetdata(const JSONRPCRequest& request)
                 "  units: (number),\n"
                 "  reissuable: (number),\n"
                 "  has_ipfs: (number),\n"
-                "  ipfs_hash: (hash) (only if has_ipfs = 1)\n"
+                "  ipfs_hash: (hash) (only if has_ipfs = 1 and that data is an ipfs hash)\n"
+                "  txid_hash: (hash) (only if has_ipfs = 1 and that data is an txid hash)\n"
```
[the remark this answers is not in the document]
Comment (blondfrogs) [reply src/rpc/assets.cpp]: Fixed

File: src/rpc/assets.cpp [hunk 10]
```diff
@@ -1169,7 +1196,8 @@ UniValue listassets(const JSONRPCRequest& request)
                 "      units: (number),\n"
                 "      reissuable: (number),\n"
                 "      has_ipfs: (number),\n"
-                "      ipfs_hash: (hash) (only if has_ipfs = 1)\n"
+                "      ipfs_hash: (hash) (only if has_ipfs = 1 and data is an ipfs hash)\n"
+                "      ipfs_hash: (hash) (only if has_ipfs = 1 and data is an txid hash)\n"
```
[the remark this answers is not in the document]
Comment (blondfrogs) [reply src/rpc/assets.cpp]: Fixed

File: src/assets/assets.cpp [hunk 11]
```diff
@@ -1120,13 +1120,13 @@ bool CReissueAsset::IsValid(std::string &strError, CAssetsCache& assetCache, boo
         }
     }
 
-    if (strIPFSHash != "" && strIPFSHash.size() != 34) {
-        strError = _("Invalid parameter: ipfs_hash must be 34 bytes.");
+    if (strIPFSHash != "" && strIPFSHash.size() != 34 && (AreMessagingDeployed() && strIPFSHash.size() != 32)) {
+        strError = _("Invalid parameter: ipfs_hash must be 34 bytes, Txid must be 64 bytes");
```
[the remark this answers is not in the document]
Comment (blondfrogs) [reply src/assets/assets.cpp]: Fixed

File: src/qt/createassetdialog.cpp (hunk 4 above)
Comment (blondfrogs) [reply src/qt/createassetdialog.cpp]: Fixed

File: src/qt/reissueassetdialog.cpp (hunk 5 above)
Comment (blondfrogs) [reply src/qt/reissueassetdialog.cpp]: Fixed

File: src/assets/assets.cpp [hunk 12]
```diff
@@ -354,13 +354,13 @@ bool CNewAsset::IsValid(std::string& strError, CAssetsCache& assetCache, bool fC
         return false;
     }
 
-    if (nHasIPFS && strIPFSHash.size() != 34) {
-        strError = _("Invalid parameter: ipfs_hash must be 34 bytes.");
+    if (nHasIPFS && strIPFSHash.size() != 34 && (AreMessagingDeployed() && strIPFSHash.size() != 32)) {
+        strError = _("Invalid parameter: ipfs_hash must be 34 bytes. Txid must be 64 bytes");
```
[the remark this answers is not in the document]
Comment (blondfrogs) [reply src/assets/assets.cpp]: Fixed

File: src/qt/createassetdialog.cpp (hunk 3 above)
Comment (blondfrogs) [reply src/qt/createassetdialog.cpp]: Fixed

**Comment by TronBlack** [comment]
Let's get this on testnet.

**Pull request merged**